### PR TITLE
adding new methods for public and private favorites

### PR DESCRIPTION
**Pull request by erqan**
implements **newfavmsg** and **getfavs** methods for adding and getting favorite posts, like retwist mechanism...

```newfavmsg <localuser> <k> <userpost-obj> [is-private] [msg]```
**is-private** if true. that favorite will be encrypted, like DMs.. 
**msg** user may want to comment while favoriting... 

```getfavs <username> <count> [{"max_id":<maxid>,"since_id":<since_id>}]```
get stored favorite posts from swarm...
returns two types of favorite, *public* and *private*..
**public favorite**
```
 {
        "isPrivate" : 0,
        "sig_userpost" : "1f1fcec966663079ae11f146f7a51ed5b4e58103b17c0576d055cad5acdb523274a13dd3ccce904293984cceef5a727f8cb48dc87c88b89ce13539fbaa81ce2e24",
        "userpost" : {
            "fav" : {
                "height" : 29159,
                "k" : 0,
                "msg" : "nice to meet you!…",
                "n" : "erkanyuksel",
                "time" : 1394914435
            },
            "height" : 85286,
            "k" : 279,
            "n" : "erqan",
            "sig_fav" : "2088a10027f134c90b0699b1d0a0795fe30b4395416cfe232a2e47b8fb49a254fa65eaaba9bf2713804f0e0f961b7aa079ebf1d2347ca187b908ba1034a73101d5",
            "time" : 1430655760
        }
    }
```
**private favorite (with getfavs)**
```
   {
        "isPrivate" : 1,
        "userpost" : {
            "fav" : {
                "height" : 85352,
                "k" : 474,
                "lastk" : 473,
                "msg" : "#twister fav(orite)s and p(rivate)fav(orite)s in swarm, like retwists.. it can be interesting!? http://i.imgur.com/8lV1ILZ.png",
                "n" : "erkanyuksel",
                "time" : 1430698498
            },
            "height" : 85409,
            "k" : 291,
            "msg" : "#deneme",
            "n" : "erqan",
            "sig_fav" : "1f1ac86561a868eee47084098a6dadb351c2fc44c7ceb7f7a443b93b16c6f4201dfb150e1ccce89c6252e54a31a9392b88f5c904f1c06dc4d4d31800dddb35bc9d",
            "time" : 1430744116
        }
    }
```
**raw encrypted object (with getpost)**
```
   {
        "sig_userpost" : "1f8252d192549fb4169ad3739fcd40bb3aeb97aa28e6d7dbc3b58c8aa4b38205da969c69e4159fd0345e79b831397f0533ad1e108cf95382f987fec7f86e0dc5c2",
        "userpost" : {
            "height" : 85409,
            "k" : 291,
            "msg" : "#deneme",
            "n" : "erqan",
            "pfav" : {
                "body" : "6c641df350538b556c1eeafe3f9e2d0bdab9882df7249c229ee77f9e8f4c9d021f590f0c6e62a974357a0d4a5ba9d0c99cdf66fbc8f27268cbf6f1bbc4db23f2aa9a71b1a53857c1705cb1fceff88108685b57ec512061c83bd86031ffa9bc6c5036980e8f8232f1eff319486d615fb1465d15cce598c9d3c55c16aa7bcdc6d67dfa2b383314536599a452b8e102ad5e633671e385e51c6eacc24fdf9702851d884295d911d3b4b5c1995fd9c0d537f1ef22db9e333bdfe08965b1c83a6a35a06fe83ae84d03d6929596ff7778ed2fe99a1054f21075ed80dff81d9398c68bf240455d58a7d930115fc10ad2f8d0c78a9b7f137c835f927b2f2ee493cf36fc990e1c54318b85376e4d634f85c8f68aa17170b986dfcb93e369130d2865deca400acbe9fb0367528c103f97c3f4448d9c",
                "key" : "03473a469e16c96f90d6230be8304d86aa1ecb8475087f2f54093a3c96fb38a7a9",
                "mac" : "98ab47414c7356b5b68af65c9976c6cb414ba2c9e7290c2798c98892bb1b4235dcda5cc0c8c7fe7eaf54e5ecaf811988a272fb711e51de78bcbc4f1bd16e9d80",
                "orig" : 302
            },
            "time" : 1430744116
        }
    }
```

### CODE REVIEW

**Comment by miguelfreitas** [comment]
Cool stuff, thanks! I've just returned from vacations so allow me a few days to review this and other pending things.

File: src/twister.h [hunk 1]
```diff
@@ -8,8 +8,12 @@
 
 #define LIBTORRENT_PORT_OFFSET 1000
 
-#define USERPOST_FLAG_RT 0x01
-#define USERPOST_FLAG_DM 0x02
+#define USERPOST_FLAG_RT    0x01
+#define USERPOST_FLAG_DM    0x02
+#define USERPOST_FLAG_FAV   0x04
+#define USERPOST_FLAG_P_FAV 0x0C
```
Comment (miguelfreitas) [inline src/twister.h]: I just found a bit funny to have a bitfield flag with two bits set at once (P_FAV is 0x04 | 0x08 = 0x0C), but ok.
The semantic is more like FAV + PRIV, but i understand that PRIV alone (0x08) doesn't make much sense...

Comment (erqan) [inline src/twister.h]: yes, that was for fetching together the **fav** and the **pfav** messages with a single *get_pieces* call in *getfavs* method.

**Comment by thedod** [comment]
Sorry for being a devil's advocate, but why add all this functionality and only settle for a single bit saying fav [or not]? Might as well be a list of tags `["#fav","#math","#lol"]` or even a dict `["#fav":1,"#math":0.8,"#lol":-0.5,"#sad":"indeed"]` (or maybe we should restrict values to 1..5 stars at 1/2 star resolution, because there are lots of algorithm for that specific range :wink:).

Twister is powerful at accountability (we can assume there are no impersonations) and persistence (you can't delete stuff), so the more expressive a single item can be, the more you can crunch intelligence deduvtions out of it "as a cloud".
For example, I can tag a twist as `#twiser_dev` (or give it "3.5 stars" in `#twister_dev` or whatever), and then you can search for `#twister_dev` and get the twists sorted by how they were tagged by others.
If I can also tag accounts, I can get twists sorted by the `#twister_dev` tags they got by *accounts* that *I* have tagged `#twister_dev`, and if tags got numeric values (e.g. "stars"), then the big data can really shine here.
No matter how deep you go with this, you can always implement a "traditional fav" with it (even with the same look and feel twitter etc. have). You implement it by tagging as `"#fav"` (or `{"#fav":5}` if it's a "star system").

Even if no one implements any of this, I can always start twisting [<140 char] json and write clients that do something funky with that :wink:

**Pull request merged**